src/_zkapauthorizer/tests/test_replicate.py

```diff
@@ -318,240 +327,284 @@ def match(self, matchee: tuple[str, Callable[[], BinaryIO]]) -> Optional[Mismatc
+        replica_dircap = grid.make_directory()
+        client = grid.client()
+
+        replica = get_tahoe_lafs_direntry_replica(client, replica_dircap)
+        service = replication_service(store._connection, replica)
+        service.startService()
+        self.addCleanup(service.stopService)
 
-        # Now both the first upload and a second upload should have completed.
-        # There is no third upload because the data for the 2nd and 3rd
-        # add_tokens calls should have been combined into a single upload.
         self.assertThat(
-            uploads,
-            MatchesListwise(
-                [
-                    match_upload(
-                        Equals("event-stream-11"),
-                        MatchesStructure(
-                            changes=HasLength(11),
-                            highest_sequence=returns(Equals(11)),
-                        ),
-                    ),
-                    match_upload(
-                        Equals("event-stream-33"),
-                        MatchesStructure(
-                            changes=HasLength(22),
-                            highest_sequence=returns(Equals(33)),
-                        ),
-                    ),
-                ],
-            ),
+            grid.list_directory(replica_dircap),
+            Contains("snapshot"),
         )
 
-        # since we've uploaded everything, there should be no
-        # events in the store
-        self.assertEqual(tuple(), get_events(tvs.store._connection).changes)
-
-    def test_snapshot(self) -> None:
+    def test_replicate(self) -> None:
         """
         Making changes to the voucher store while replication is turned on
         causes event-stream objects to be uploaded.
         """
+        tvs = self.useFixture(TemporaryVoucherStore(aware_now))
 
-        def get_config(rootpath, portnumfile):
-            return EmptyConfig(FilePath(rootpath))
-
-        tvs = self.useFixture(TemporaryVoucherStore(get_config, lambda: aware_now()))
-
+        grid = MemoryGrid()
+        replica_cap = grid.make_directory()
         rwcap_file = FilePath(tvs.config.get_private_path(REPLICA_RWCAP_BASENAME))
         rwcap_file.parent().makedirs()
-        rwcap_file.setContent(b"URL:DIR2:stuff")
+        rwcap_file.setContent(replica_cap.encode("ascii"))
+
+        # Predicate to check if the replica directory has at least some number
+        # of files in it.
+        has_files_bound = partial(has_files, grid, replica_cap)
 
         # we use this to contol when the first upload happens, so that we
         # actually use the queue
-        wait_d: Deferred[None] = Deferred()
-
-        uploads = []
-        upload_completed = False
-
-        async def uploader(name, get_data):
-            uploads.append((name, get_data))
-            await wait_d
-            nonlocal upload_completed
-            upload_completed = True
+        delay_controller, delay_client = delayedProxy(
```

Comment:
Using the `delay_controller` to exercise the precise cases we want to exercise in the tests sometimes works out nicely but sometimes ends up pretty gross, with some hard-coded number of `run` calls required to make sure the internal state ends up in just the right place.

I think an improvement would be to write these as Hypothesis stateful tests where "an async thing you started has completed" is one possible transition for the state machine.  This would let Hypothesis explore the stateful space for us in one general test instead of requiring us to hand-craft careful transitions to just the right case we were interested in for one specific test.

Reply:
My attempt at writing a hypothesis stateful-thing for magic-folder failed .. so either I don't understand their docs yet or tried something too ambitious :)

Reply:
I guess another disadvantage of the "delay_controller" is that if the test would "normally" fail, it'll just hang now? (because e.g. 3 files never appear)

Reply:
> I guess another disadvantage of the "delay_controller" is that if the test would "normally" fail, it'll just hang now? (because e.g. 3 files never appear)

Potentially, yea. :(  `delay_controller.run` mitigates this somewhat by raising an exception if there is nothing delayed - so if the failure is that the implementation stops trying to do anything before getting to the desired state, it'll fail quickly.  But if the failure is that the implementation keeps trying to do work but never gets to the desired state, yea, repeat forever.

This could be mitigated by adding a constraint on the maximum number of repeats.  I might prefer instead to spend the effort trying to write the tests so they don't have these sensitivities on implementation details though.

Reply:
> My attempt at writing a hypothesis stateful-thing for magic-folder failed .. so either I don't understand their docs yet or tried something too ambitious :)

I _hope_ this will be possible for the replication service but I'm not completely sure I already know how to do it.  So we'll see...